scripts: Scripts to build a relational database

teburd opened the pull request:
A quick hack up of ideas built up from the last few days, and a nice brain storm session with @nashif and @galak, after discussions with @mbolivar on how the resulting device tree comes about.

The first step in being able to easily query does a board support a particular driver API is getting the list of compatibles a board is related to.

Adds a few scripts to generate a sqlite3 db with an initial table containing the pair of (board, compatible) strings.

This can be used to then query for example, which boards require which compatibles, and vice versa.

Additional properties could be added to the table if needed for further filtering.

A second table generated from Kconfig which has the tuple (compatible, api) would enable relating boards to Zephyr's HAL APIs.

This doesn't cover all relationships to all APIs in Zephyr, but at least a large set of them is. E.g. sys timer, log backends, and probably others aren't.

This database could also be used to generate/validate a board doc table or board yaml file describing what is supported, by which drivers, and so on.

If this database were available prior to running twister or ci, perhaps the data could be used to decide which tests to build/run on which boards.

The database takes a few minutes to build for me which is pretty reasonable. xargs could be replaced with gnu parallel and probably go faster.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
scripts: Scripts to build a device database

The first step in being able to easily query does a board support
a particular driver API is getting the list of compatibles a board
is related to.

Adds a few scripts to generate a sqlite3 db with an initial table
containing the pair of (board, compatible) strings.

This can be used to then query for example, which boards require
which compatibles, and vice versa.

Additional properties could be added to the table if needed for
further filtering.

A second table generated from Kconfig which has the tuple (compatible, api)
such that the final connection lto which boards support which device APIs
from the device tree and Kconfig depends on/selects is possible.

This could also be used to generate and validate a board doc table or
board yaml file describing what is supported, by which drivers, and so on.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

mbolivar-nordic reviewed (commented):
I see where something like this could be useful based on our discussion.

nashif commented:
@teburd  thank, started playing with this, and with this data we can already updating out of sync information we have in the tree, for example board yaml files, for example, updating the `supported` field to allow better filtering of tests:

```patch
<cut>
--- a/boards/arm/96b_avenger96/96b_avenger96.yaml
+++ b/boards/arm/96b_avenger96/96b_avenger96.yaml
@@ -9,6 +9,12 @@ toolchain:
 supported:
   - gpio
   - shell
+  - spi
+  - i2c
+  - watchdog
+  - uart
+  - led
+  - dma
 testing:
   ignore_tags:
     - cmsis_rtos_v2
diff --git a/boards/arm/96b_carbon/96b_carbon.yaml b/boards/arm/96b_carbon/96b_carbon.yaml
index 53bb4aa953..2f24640e57 100644
--- a/boards/arm/96b_carbon/96b_carbon.yaml
+++ b/boards/arm/96b_carbon/96b_carbon.yaml
@@ -13,5 +13,13 @@ supported:
   - counter
   - spi
   - usb_device
+  - watchdog
+  - rtc
+  - pwm
+  - adc
+  - uart
+  - led
+  - flash
+  - dma
 ram: 96
 flash: 512
diff --git a/boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.yaml b/boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.yaml
index e83f8cdf14..9adee4e38a 100644
--- a/boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.yaml
+++ b/boards/arm/96b_carbon_nrf51/96b_carbon_nrf51.yaml
@@ -10,3 +10,10 @@ toolchain:
   - xtools
 supported:
   - ble
+  - spi
+  - gpio
+  - rtc
+  - pwm
+  - adc
+  - uart
+  - flash
</cut>
```

need to see how this plays out in real life.

Commit:
scripts: Sqlite from twister output script

Rather than trying to use shell scripting to do this, which requires
spawning lots of python scripts which takes time... use pythons builtin
concurrent module to do things concurrently.

Take a twister output directory and look for board, test pairs, reading in
generated metadata files such as edt.pickle and compile_commands into
sqlite tables which can then be used to query things like which test
depends on which files.

While compile_commands.json is limited in its scope a better generated
file using some cmake hackery seems plausible.

Imagine several tables exist... which can then be joined on to determine
prior to running cmake which tests may need to be built.

A table containing (board, test, config) tuples and another containing
(config, file) tuples. A join can then be done to determine which (board,
test) pairs are needed given a set of files that have been mutated.

Commit:
scripts: build sqlite

Adds kconfig parsing, building a graph of symbol dependencies
and automatic selections.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
Inserts kconfig symbol selects/depends on

Makes querying the Kconfig symbol graph somewhat nice though...
many symbols are missing while parsing Kconfig the way the check
contribution script works. This is unfortunate, to get the full set
of symbols I'd need to I think run cmake for every board and test pair
parsing the output, as there are symbols generated by devicetree.

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>

Commit:
Pausing on this, but it does make a little db now

Signed-off-by: Tom Burdick <thomas.burdick@intel.com>
